File: android/app/src/main/java/com/bitcraftapps/stax/ui/composables/UpgradeBanner.kt
package com.bitcraftapps.stax.ui.composables

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.bitcraftapps.stax.data.billing.Feature
import com.bitcraftapps.stax.ui.theme.StaxPrimary

@Composable
fun UpgradeBanner(
    message: String,
    onUpgrade: () -> Unit,
    modifier: Modifier = Modifier
) {
    Card(
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceContainerHigh
        ),
        modifier = modifier.fillMaxWidth()
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Default.Lock,
                contentDescription = null,
                modifier = Modifier.size(20.dp),
                tint = StaxPrimary
            )
            Spacer(modifier = Modifier.width(12.dp))
            Text(
                text = message,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.weight(1f)
            )
            Spacer(modifier = Modifier.width(12.dp))
            Button(
                onClick = onUpgrade,
                colors = ButtonDefaults.buttonColors(
                    containerColor = StaxPrimary
                ),
                shape = MaterialTheme.shapes.small
            ) {
                Text(
                    "Upgrade",
                    style = MaterialTheme.typography.labelMedium
                )
            }
        }
    }
}

@Composable
fun UpgradeDialog(
    feature: Feature,
    onUpgrade: () -> Unit,
    onDismiss: () -> Unit
) {
    val (headline, description) = when (feature) {
        Feature.SCAN -> "Daily Scan Limit Reached" to
                "You've used all 5 free scans for today. Upgrade to Premium for unlimited scans every day."
        Feature.AI_SCAN -> "Premium Feature" to
                "AI Stack Counter uses cloud vision to count your chips instantly. Available with STAX Premium."
        Feature.SESSION_CREATE -> "Session Limit Reached" to
                "Free accounts can track up to 3 sessions. Upgrade to Premium for unlimited sessions."
        Feature.PHOTO_ADD -> "Photo Limit Reached" to
                "Free accounts can add up to 10 photos per session. Upgrade to Premium for unlimited photos."
        Feature.CHIP_CONFIG -> "Premium Feature" to
                "Customize chip colors and values for multiple casinos with STAX Premium."
        Feature.FAVORITES -> "Favorites Limit Reached" to
                "Free accounts can save up to 3 favorite card rooms. Upgrade for unlimited favorites."
        Feature.SHARE_CLEAN -> "Premium Feature" to
                "Export clean, watermark-free images with STAX Premium."
    }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                headline,
                style = MaterialTheme.typography.headlineSmall
            )
        },
        text = {
            Text(
                description,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        },
        confirmButton = {
            Button(
                onClick = onUpgrade,
                colors = ButtonDefaults.buttonColors(containerColor = StaxPrimary)
            ) {
                Text("Upgrade")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("Maybe Later")
            }
        }
    )
}

File: ios/Stax/Components/UpgradeBanner.swift
import SwiftUI

// MARK: – Inline upgrade banner

struct UpgradeBanner: View {
    let message: String
    let onUpgrade: () -> Void

    @State private var dismissed = false

    var body: some View {
        if !dismissed {
            HStack(spacing: 12) {
                Image(systemName: "lock.fill")
                    .foregroundColor(.staxPrimary)
                    .font(.subheadline)

                Text(message)
                    .font(.caption)
                    .foregroundColor(.white)
                    .frame(maxWidth: .infinity, alignment: .leading)

                Button("Upgrade") {
                    onUpgrade()
                }
                .font(.caption.bold())
                .foregroundColor(.white)
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(Color.staxPrimary)
                .cornerRadius(8)

                Button {
                    withAnimation { dismissed = true }
                } label: {
                    Image(systemName: "xmark")
                        .font(.caption)
                        .foregroundColor(.staxOnSurfaceVar)
                }
            }
            .padding(12)
            .background(Color.staxSurfaceHigh)
            .cornerRadius(14)
            .overlay(
                RoundedRectangle(cornerRadius: 14)
                    .stroke(Color.staxPrimary.opacity(0.35), lineWidth: 1)
            )
        }
    }
}

// MARK: – Session usage progress bar for free users

struct SessionUsageBar: View {
    let used: Int
    let limit: Int
    let onUpgrade: () -> Void

    private var atLimit: Bool { used >= limit }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 0) {
                Image(systemName: atLimit ? "exclamationmark.circle.fill" : "chart.bar.fill")
                    .font(.caption)
                    .foregroundColor(atLimit ? .staxLoss : .staxPrimary)
                    .padding(.trailing, 6)
                Text(atLimit ? "Session limit reached" : "Free plan · \(used) / \(limit) sessions used")
                    .font(.caption.bold())
                    .foregroundColor(atLimit ? .staxLoss : .white)
                Spacer()
                Button("Upgrade") { onUpgrade() }
                    .font(.caption.bold())
                    .foregroundColor(.white)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 5)
                    .background(Color.staxPrimary)
                    .cornerRadius(8)
            }
            ProgressView(value: Double(min(used, limit)), total: Double(limit))
                .tint(atLimit ? .staxLoss : .staxPrimary)
                .scaleEffect(x: 1, y: 1.6, anchor: .center)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
        .background(Color.staxSurfaceHigh)
        .cornerRadius(14)
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(atLimit ? Color.staxLoss.opacity(0.35) : Color.staxPrimary.opacity(0.25), lineWidth: 1)
        )
    }
}

// MARK: – Feature-specific upgrade alert

struct UpgradeAlert: View {
    let feature: Feature
    let onUpgrade: () -> Void
    let onDismiss: () -> Void

    var body: some View {
        VStack(spacing: 20) {
            Image(systemName: "lock.fill")
                .font(.system(size: 40))
                .foregroundColor(.staxPrimary)

            VStack(spacing: 8) {
                Text(headline)
                    .font(.title3.bold())
                    .foregroundColor(.white)
                    .multilineTextAlignment(.center)

                Text(description)
                    .font(.subheadline)
                    .foregroundColor(.staxOnSurfaceVar)
                    .multilineTextAlignment(.center)
            }

            VStack(spacing: 10) {
                Button("Upgrade to Premium") {
                    onUpgrade()
                }
                .font(.headline)
                .foregroundColor(.white)
                .frame(maxWidth: .infinity)
                .padding(.vertical, 14)
                .background(Color.staxPrimary)
                .cornerRadius(14)

                Button("Maybe Later") {
                    onDismiss()
                }
                .font(.subheadline)
                .foregroundColor(.staxOnSurfaceVar)
            }
        }
        .padding(28)
        .background(Color.staxSurface)
        .cornerRadius(22)
        .shadow(color: .black.opacity(0.5), radius: 20)
        .padding(.horizontal, 24)
    }

    private var headline: String {
        switch feature {
        case .scan:        return "Daily Scan Limit Reached"
        case .aiScan:      return "AI Scan is Premium"
        case .sessionCreate: return "Session Limit Reached"
        case .photoAdd:    return "Photo Limit Reached"
        case .chipConfig:  return "Multiple Casino Config is Premium"
        case .favorites:   return "Favorites Limit Reached"
        case .shareClean:  return "Clean Export is Premium"
        }
    }

    private var description: String {
        switch feature {
        case .scan:        return "You've used all \(FreeTierLimits.maxScansPerDay) free scans for today. Upgrade for unlimited scanning."
        case .aiScan:      return "The AI Stack Counter is available exclusively to Premium subscribers."
        case .sessionCreate: return "Free accounts can track up to \(FreeTierLimits.maxSessions) sessions. Upgrade for unlimited sessions."
        case .photoAdd:    return "Free accounts can add up to \(FreeTierLimits.maxPhotosPerSession) photos per session. Upgrade for unlimited photos."
        case .chipConfig:  return "Configure chips for all casinos with a Premium subscription."
        case .favorites:   return "Free accounts can save up to \(FreeTierLimits.maxFavorites) favorites. Upgrade for unlimited favorites."
        case .shareClean:  return "Export clean images without watermarks with a Premium subscription."
        }
    }
}
